src/psyclone/transformations.py

```diff
@@ -2663,39 +2655,87 @@ def apply(self, node, options=None):
-                                     a GOInvokeSchedule.
-        :raises TransformationError: if there are no Loops within the \
-                                     proposed region.
+        :raises NotImplementedError: if the supplied Nodes belong to
+            a GOInvokeSchedule.
+        :raises TransformationError: if there is an access to an assumed-size
+            character variable within the region.
+        :raises TransformationError: if the proposed region contains a call to
+            an intrinsic that is not available on the accelerator.
+        :raises TransformationError: if there are no Loops within the
+            proposed region and options["disable_loop_check"] is not True.
 
         '''
         # Ensure we are always working with a list of nodes, even if only
         # one was supplied via the `nodes` argument.
         node_list = self.get_node_list(nodes)
 
         # Check that the front-end is valid
-        sched = node_list[0].ancestor((NemoInvokeSchedule,
-                                       LFRicInvokeSchedule))
-        if not sched:
+        if node_list[0].ancestor(GOInvokeSchedule):
             raise NotImplementedError(
-                "OpenACC kernels regions are currently only supported for the "
-                "Nemo and LFRic InvokeSchedules")
+                "OpenACC kernels regions are not currently supported for "
+                "GOcean InvokeSchedules")
         super().validate(node_list, options)
 
+        # The regex we use to determine whether a character declaration is
+        # of assumed size ('LEN=*' or '*(*)').
+        assumed_size = re.compile(r"\(\s*len\s*=\s*\*\s*\)|\*\s*\(\s*\*\s*\)")
+
+        # Check that there are no assumed-size character variables as these
+        # causes an Internal Compiler Error with NVHPC.
+        for node in node_list:
+            for lit in node.walk(Literal):
+                if lit.datatype.intrinsic == ScalarType.Intrinsic.CHARACTER:
+                    # We've found a character literal so we go up to the
+                    # ancestor statement and then check the types of all
+                    # symbols that are referenced by it.
+                    stmt = lit.ancestor(Statement)
+                    for ref in stmt.walk(Reference):
+                        if not ref.symbol.is_argument:
+                            # Only arguments can be of assumed length.
+                            continue
+                        # We only need to check the datatype of the underlying
+                        # Symbol.
+                        dtype = ref.symbol.datatype
+                        # Currently the fparser2 frontend does not support any
+                        # type of LEN= specification on a character variable so
+                        # we resort to a regex to check whether it is assumed-
+                        # size.
+                        if isinstance(dtype, UnsupportedFortranType):
```

Comment:
Accessing *any* UnsupportedType inside a kernels is kind of dangerous, would `if isinstance(dtype, UnsupportedType): raise TransformationError` reduce the performance?

What happens with assumed-size arrays that are accessed from a module?
`use other, only: my_assumed_size_char`

What happens if there is no Literal? Are this ok on the GPU or just not spotted?
`if (my_var == my_var2)` and one of this is assumed size.

No need to change the implementation if any of these 3 cases causes problems. We just need to document well the current limits of this validation and it would be good to have a test for each of this, regardless of what the behaviour is, to keep track of what we do.

Reply:
> What happens with assumed-size arrays that are accessed from a module?
> `use other, only: my_assumed_size_char`

This one is easy - they can only be subroutine arguments.

Reply:
> What happens if there is no Literal? Are this ok on the GPU or just not spotted?
> `if (my_var == my_var2)` and one of this is assumed size.

We wouldn't spot this one. I don't know if that would cause a problem but I will test.

Reply:
It will cause a problem. Even with nvhpc 24.5 the following code gives an ICE:

```
  subroutine my_sub(arg, my_var)

    character*(*) :: arg
    character(len=10) :: local, another
    real*8, dimension(:) :: my_var
    another = "andy"
    !$acc kernels
    local = arg
    if(local == another)then
       my_var(:) = 0.0d0
    end if
    !$acc end kernels
```
gives:
```
]$ nvfortran -c -acc assumed_size_char.f90
NVFORTRAN-S-0000-Internal compiler error. size_of: attempt to size assumed size character       0  (assumed_size_char.f90: 12)
NVFORTRAN-S-0000-Internal compiler error. size_of: attempt to size assumed size character       0  (assumed_size_char.f90: 17)
  0 inform,   0 warnings,   2 severes, 0 fatal for my_sub
```

Reply:
In fact, as illustrated by the example above, the expression inside the ACC region doesn't actually have to directly involve the assumed-size variable at all. However, if I hoist the assignment that uses it (`local = arg`) out of the kernels region then the ICE goes away. Therefore, the rule seems to be that if there's a reference to an assumed-size character array inside a Kernels region then there will probably be trouble.

Reply:
I've re-written the algorithm to do this check now. It should be both more efficient and will catch this particular case.